doc/sphinx/language/core/conversion.rst

```diff
@@ -3,11 +3,75 @@
 Conversion rules
 ----------------
 
-|CiC| has conversion rules that can be used to determine if two
-terms are equal by definition, or :term:`convertible`.
-Conversion rules consist of reduction rules and expansion rules.
-See :ref:`applyingconversionrules`,
-which describes tactics that apply these conversion rules.
+Conversion rules convert terms to other terms that are equivalent by
+definition under the rules of |CiC|.  Conversion rules can be used, for example,
+to determine if two terms are equal by definition, or :term:`convertible`, by
+converting both terms to a normal form, then verifying they are syntactically
+equal (ignoring differences in the names of bound variables
+:term:`alpha-convertibility <alpha-convertible>`).
+
+Conversion rules can be reductions or expansions.
+Reductions convert terms to something that is incrementally closer to its
+normal form.  For example, :term:`zeta-reduction` removes
+:n:`let @ident := @term__1 in @term__2` constructs from a term by replacing
+:n:`@ident` with :n:`@term__1` wherever :n:`@ident` appears in :n:`@term__2`.
+The resulting term may be longer or shorter than the original.
+
+.. coqtop:: all
+
+   Eval cbv zeta in let i := 1 in i + i.
+
+Expansions are reductions applied in the opposite direction,
+for example expanding `2 + 2` to `let i := 2 in i + i`.  While applying
+reductions (except for eta-reduction) gives a unique result , the associated
+expansion may not give a unique result.  For example, `2 + 2` could also be
+expanded to `let i := 2 in i + 2`.  Expansions that give a unique
+result are also referred to as *contractions*.
+
+The normal form is defined as the result of applying a particular
+set of conversion rules (beta-, delta-, iota- and zeta-reduction and eta-expansion)
+repeatedly until it's no longer possible to apply any of them.
+
+Some reductions can be considered computations, for example reducing `2*3+4` with
+`cbv beta delta iota` to `10`, which requires applying several reduction rules
+repeatedly.
+
+The useful conversion rules are:
+
+   .. list-table::
+      :header-rows: 1
+
+      * - Conversion name
+        - Description
+
+      * - beta-reduction
+        - eliminates `fun`
+
+      * - delta-reduction
+        - replaces a defined symbol with its definition
+
+      * - cofix-reduction
+        - replace a `cofix` symbol with its definition
+
+      * - fix-reduction
+        - replaces a `fix` symbol with its definition
+
+      * - match-reduction
+        - eliminates `match`
+
+      * - iota-reduction
+        - match-, fix- and cofix-reduction together
+
+      * - zeta-reduction
+        - eliminates `let`
+
+      * - eta-expansion
+        - TBA
```

Comment:
Replaces a term `f` of type `forall a : A, B` with `fun x : A => f x`

Reply:
Hard to create a convincing example in coqtop.  Could I have something vaguely like `forall a : A, B = fun x : A => f` which `cbv` fails to convert (lacking eta) while, say, `reflexivity` includes eta and so show they are equivalent?

Reply:
The canonical example is the one I gave in the issue tracker involving composition of a variable with the identity function.  I think that should be the example that gets included.

Note also that there is both function eta and record eta, and the latter is controlled with Set Primitive Projections.  Record eta contraction is turning (fst x, snd x) into x and similar for other records.

Reply:
BTW, I haven't studied the typing rules much yet, so I don't know how to construct a term that has the type `forall a : A, B`.

Reply:
> The canonical example is the one I gave in the issue tracker involving composition of a variable with the identity function ...

Which PR?  I think not #2860, whichi is eta-reduction rather than eta-expansion.

Reply:
> Which PR? I think not #2860, which is eta-reduction rather than eta-expansion.

In https://github.com/coq/coq/issues/13811#issuecomment-771104374.  Note that while "reduction" is directed (towards normalization), supporting a reduction rule in conversion necessarily means supporting both directions, because "convertible" is an equivalence relation.

The reason eta-expansion is useful is because putting everything in eta-long form (fully eta-expanded) does not break typing and allows untyped conversion to support eta.  eta-contraction, on the other hand, is much harder to implemented when you're not carrying around types.  It's primarily useful for displaying things in a nicer format (I imagine that `simpl` would want to do eta-contraction, while `cbv` would not), and occasionally for de-duplicating work in a non-shared reduction strategy (computing `(fst x, snd x)` where `x` is a complicated term may compute `x` twice rather than once).

Reply:
> BTW, I haven't studied the typing rules much yet, so I don't know how to construct a term that has the type `forall a : A, B`.

You can access to one by writing `fun f : (forall a : A, B) => ...`.  Or via `Axiom`.  Alternatively, if you want a concrete term for some particular choice of `A` and `B`, `Nat.add` is one, and `id` (or `fun a => a`) is another.  Really I'm just talking about any term which is a function i.e., which can take arguments / be applied to things.

Reply:
AFAIK there's no tactic that generates and displays the eta-expansion of a term, so not sure it makes sense for me to try give an example--hard for me to know if it's correct.